content/tutorials/reactor-event-aws.textile

```diff
@@ -0,0 +1,785 @@
+// -----------------------------------------
+const ablyApiKey = 'YOUR_API_KEY_GOES_HERE';
+// -----------------------------------------
+
+// This is the handler function that AWS will invoke:
+exports.handler = (event, context, callback) => {
+  
+  // Get the Lex runtime from the AWS client library:
+  const lex = new AWS.LexRuntime();
+
+  // Extract the customer message and ID from the data received from Ably:
+  const data = JSON.parse(event.messages[0].data);
+
+  // `data.user` is the customer ID from our postMessage function:
+  const channel = 'pizza:bot:' + data.user;
+
+  // Construct a standard Node.js-style callback function that will be invoked
+  // when our call AWS Lex completes:
+  const onResponse = (err, result) => err
+    ? callback(err)
+    : postMessage(ablyApiKey, channel, result.message, callback);
+
+  // Below, in addition the customer's message, we include their ID, because Lex
+  // can use it to "remember" the conversation for a little while, which allows
+  // the customer to reply with short messages that refer contextually to
+  // previous responses emitted by the bot.
+  lex.postText({
+    botName: 'PizzaAssistant',
+    botAlias: 'PizzaAssistant',
+    userId: data.user,
+    inputText: data.message
+  }, onResponse);
+};
+
+// The onResponse callback function, defined above, calls postMessage in order
+// to dispatch the response back to Ably. The Ably client API isn't available on
+// AWS, at least not without a bunch of extra setup. Instead, the postMessage
+// function will simply make use of Ably's REST API.
+
+function postMessage (apiKey, channel, message, callback) {
+  // Prepare the data to post to Ably:
+  const data = JSON.stringify({
+    name: 'bot',
+    data: message
+  });
+
+  // HTTP configuration that includes the bot response channel and your API key,
+  // included as a base64-encoded authorization header:
+  const options = {
+    host: 'rest.ably.io',
+    port: 443,
+    path: `/channels/${channel}/messages`,
+    method: 'POST',
+    headers: {
+      'Authorization': `Basic ${new Buffer(apiKey).toString('base64')}`,
+      'Content-Type': 'application/json',
+      'Content-Length': Buffer.byteLength(data)
+    }
+  };
+
+  // Construct the HTTPS request using the above configuration data:
+  const req = https.request(options, (res) => {
+    let output = '';
+    res.setEncoding('utf8');
+    res.on('data', (chunk) => output += chunk);
+    res.on('end', () => (statusCode, result) => {
+      // We should still use the Lambda function's default callback, in case
+      // there are any HTTPS errors returned by the Ably API. We can use further
+      // AWS tools for investigation, if this is the case.
+      callback(null, { statusCode, result });
+    });
+  });
+
+  // As above, but trap errors that prevent the call from happening at all:
+  req.on('error', (err) => {
+    callback(err);
+  });
+
+  // Send the payload, and we're done:
+  req.write(data);
+  req.end();
+}
+```
```

Comment:
Sure, but it seems the user would have to create a serverless deploy package for this (correct me if I'm wrong on that account). The tutorial just uses the AWS UI for the tutorial, so the code uses the REST API on that basis.

Reply:
Note that I made the following comment on the Slack channel:

> It may end up being worthwhile updating the main reactor event AWS tutorial to use the new chat bot script I wrote, instead of Lex (seeing as the new script is a lot more robust). I'd prefer to do this as a separate job though, if you decide that's what you want.

As an extension to the above, I could re-focus the tutorial on using a serverless deploy package instead.

Reply:
> Sure, but it seems the user would have to create a serverless deploy package for this (correct me if I'm wrong on that account). The tutorial just uses the AWS UI for the tutorial, so the code manually calls the REST API on that basis

Thing is, we want to seriously discourage people from using the REST api manually if at all possible, in favour of the client libs, for a number of reasons (e.g. automatic retries to fallback hosts to route around region-specific or DNS issues). As such we would like all the tutorials to use the official client libs, even if it's a bit of extra work.

You do have to create a package, but you only need a pretty minimal one. I was actually thinking of making an official example lambda package anyway. Tell you what -- I'll do that now, so you can link to it in the tutorial, to save the trouble of explaining how to create the package from scratch. so something "Clone this example package [to be done], replace the contents of index.js with the following code, and run `npm run zip` to generate the zip package".

Reply:
Example lambda package up at https://github.com/ably/example-lambda-function